### PR TITLE
[IPU] support dy2static for IPU

### PR DESCRIPTION
**Pull request by czr-gc**
### PR types
New features

### PR changes
Others

### Describe

增加IPU上动转静的支持。IPU动转静借助paddle动转静到实现，区别在于：
- 在利用paddle动转静机制生成静态图后，IPU需要多一步操作，将静态图进一步转换成IPU上的静态图。
- IPU上的forward/backward均在一次session调用过程中完成，所以无法支持譬如`loss.backward()`, `optimizer.step()`等反向接口。因此新增了`identity_loss`op用于在IPU上动态标记loss。

### CODE REVIEW

**Review by qili93 — DISMISSED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            return x, loss
+        return x
+
+
+class TestBase(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        paddle.disable_static()
+        cls.save_path = os.path.join(os.getcwd(), "temp_save")
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py]: Recommend to use `tempfile`,for example ( see test_grad.py)
```python
import tempfile

class TestGradLinear(TestGrad):

    def setUp(self):

        self.temp_dir = tempfile.TemporaryDirectory()
        self.infer_model_path = os.path.join(self.temp_dir.name,
                                             'double_grad_infer_model')
        self.train_model_path = os.path.join(self.temp_dir.name,
                                             'double_grad_train_model')

    def tearDown(self):
        self.temp_dir.cleanup()
```

File: python/paddle/fluid/compiler.py [hunk 2]
```diff
@@ -499,6 +499,175 @@ def _get_places(self, place, place_list):
...
+        def append_backward_desc():
+            program = concrete_program.main_program
+
+            # backward with optimizer to add backward graph to program
+            backward.gradients_with_optimizer(program, ipu_strategy._optimizer)
+
+            # initialize backward parameters
+            exe = paddle.static.Executor(paddle.CPUPlace())
+            startup_program = paddle.static.default_startup_program()
+            exe.run(startup_program)
```
Comment (Aurelius84) [inline python/paddle/fluid/compiler.py:541]: Why we need execute startup_program ? Is it necessary?

File: python/paddle/fluid/compiler.py [hunk 3]
```diff
@@ -535,10 +704,39 @@ def __init__(self):
...
             self.custom_op_names = []
+            self.need_compile = True
         else:
             raise RuntimeError(
                 "Can not use IpuStrategy in non IPU compiled environment, please re-compile with WITH_IPU=ON."
             )
+        from paddle import in_dynamic_mode
+        if in_dynamic_mode():
+            IpuDynamicPatcher.patch_program_cache(self)
+            IpuDynamicPatcher.patch_lr_scheduler(self)
```
Comment (Aurelius84) [inline python/paddle/fluid/compiler.py]: Here we replace `__getitem__` of ProgramCache, whether we need patch this method back into original one ? I mean if you replace it here and we want to use @to_static normally somewhere at same time, does it raise error ?

File: python/paddle/fluid/compiler.py [hunk 4]
```diff
@@ -499,6 +499,175 @@ def _get_places(self, place, place_list):
...
+                                 concrete_program,
+                                 class_instance=None):
+        """
+        Convert the ConcreteProgram to IPUConcreteProgram.
+        """
+        from ..fluid.dygraph.base import switch_to_static_graph
+        from ..fluid import backward
+        from ..fluid.initializer import Constant
+        from ..fluid.framework import device_guard
+        import paddle
```
Comment (Aurelius84) [inline python/paddle/fluid/compiler.py:523]: Recommend to add import statement in front of file.

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py (hunk 1 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py]: Thanks, will update it.

File: python/paddle/fluid/compiler.py (hunk 2 above)
Comment (czr-gc) [reply python/paddle/fluid/compiler.py:541]: Yes, startup_program is to initialize the backward related paramters. 

IPU executes the forward&backward in oneshot and doesn't support backward api like `loss.backward()`. So we need to create backward graph and init paramters here.

File: python/paddle/fluid/compiler.py (hunk 3 above)
Comment (czr-gc) [reply python/paddle/fluid/compiler.py]: The main idea here is to convert the paddle program to IPU-supported program. And yes, it may introduce the error if user want to use `to_static` normally somewhere.

Will consider an registration here instead of direct patch.

File: python/paddle/fluid/compiler.py (hunk 4 above)
Comment (czr-gc) [reply python/paddle/fluid/compiler.py:523]: I was considering those imports are only for IPU usage and don't want to import those modules for normally uses. I can move it to beginning of file if you think it's better.

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py (hunk 1 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_fp16_ipu.py]: Updated.

File: python/paddle/fluid/compiler.py (hunk 3 above)
Comment (czr-gc) [reply python/paddle/fluid/compiler.py]: Added apis to register/release the patch.

File: python/paddle/fluid/compiler.py (hunk 4 above)
Comment (czr-gc) [reply python/paddle/fluid/compiler.py:523]: unfortunately, move those imports at front introduced import loop issue. We may need to keep it here.

**Review by Aurelius84 — DISMISSED** [review]

**Comment by gglin001** [comment]
@qili93  hi, 请问 这个 PR 还需要更多 的修改/review 吗? 目前必要的 CI 都已经ready了, 

`PR-CI-Kunlun-KP-Build` 的失败看起来和修改没有关系

**Review by Aurelius84 — DISMISSED** [review]
LGTM

**Comment by czr-gc** [comment]
@qili93 按意见增加了相关接口的文档，以及pr： https://github.com/PaddlePaddle/docs/pull/4932

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_ipu.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        if use_ipu:
+            ipu_strategy.release_patch()
+
+        return np.array(result)
+
+    def test_training(self):
+        ipu_loss = self._test(True).flatten()
+        cpu_loss = self._test(False).flatten()
+
+        self.assertTrue(np.allclose(ipu_loss, cpu_loss, atol=1e-4))
```
Comment (qili93) [inline python/paddle/fluid/tests/unittests/ipu/test_dy2static_ipu.py:100]: 确认下这个单测就是跑了动态图，并没有使用到动转静的功能对不？没有看到代码里面有对于 paddle.jit.to_static的调用

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            result.append(loss)
+
+        if use_ipu:
+            paddle.fluid.core.IpuBackend.get_instance().weights_to_host()
+
+        paddle.save(model.state_dict(), model_path)
+        paddle.save(optim.state_dict(), optim_path)
+
+        model.set_state_dict(paddle.load(model_path))
+        optim.set_state_dict(paddle.load(optim_path))
```
Comment (qili93) [inline python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: paddle.save主要是动态图用于保存模型的参数，建议这里参考 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py 中 check_jit_save_load 的 case 设计。

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            return x, loss
+        return x
+
+
+class TestBase(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        paddle.disable_static()
+        cls.save_path = os.path.join(os.getcwd(), "temp_save")
```
Comment (qili93) [inline python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: 为了避免IO冲突，目前Paddle对存在文件读写的单测全部采用tempfile，请参考 https://github.com/PaddlePaddle/Paddle/pull/43387 这个PR中的修改，改用 tempfile.TemporaryDirectory() 的形式

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    def test_training(self):
+        cpu_loss = self._test(False).flatten()
+        ipu_loss = self._test(True).flatten()
+
+        self.assertTrue(np.allclose(ipu_loss, cpu_loss, atol=1e-4))
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (qili93) [inline python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: 建议参考 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py 中的代码 (1) 增加单测case (2) 合并 test_dy2static_ipu.py 和 test_dy2static_save_load_ipu.py 两个文件。

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_ipu.py (hunk 5 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_ipu.py:100]: 在网络前向有定义装饰器：
```
    @to_static()
    def forward(self, x, target=None):
```

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py (hunk 6 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: `jit.save`和`jit.load`后的模型会被转换为`TranslatedLayer`, 在IPU上适配比较麻烦，所以目前采用的是动态图直接保存模型参数，并加载到动态图上的方式。

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py (hunk 7 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: 已更改

File: python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py (hunk 8 above)
Comment (czr-gc) [reply python/paddle/fluid/tests/unittests/ipu/test_dy2static_save_load_ipu.py]: 测例已经合并到test_dy2static_ipu.py

**Review by qili93 — DISMISSED** [review]
LGTM

**Comment by gglin001** [comment]
CI `PR-CI-pre-commit` 的失败是符合预期的, PR https://github.com/PaddlePaddle/Paddle/pull/43770  增加了一笔提交用来修改格式(如果应用到这个PR上将使得代码 review 很困难), code format 也许可以在这个PR合入后在新的PR里面解决
CI `PR-CI-Static-Check` 的失败也是因为 code format 的问题
